File: MiniProject01/MiniProject01/MovieTableViewController.m
//
//  MovieTableViewController.m
//  MiniProject01
//
//  Created by 김성준 on 2016. 11. 15..
//  Copyright © 2016년 poppay. All rights reserved.
//

#import "MovieTableViewController.h"
#import "MovieDetailViewController.h"
#import "SWRevealViewController.h"
#import "MovieCell.h"
#import "MovieVO.h"

@interface MovieTableViewController ()

@end

@implementation MovieTableViewController{
    MovieVO *mvo;
    NSMutableArray<MovieVO *> *list;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    SWRevealViewController *revealViewController = self.revealViewController;
    if ( revealViewController )
    {
        [self.sidebarButton setTarget: self.revealViewController];
        [self.sidebarButton setAction: @selector( revealToggle: )];
        [self.view addGestureRecognizer:self.revealViewController.panGestureRecognizer];
    }
    
    list = [NSMutableArray<MovieVO *> new];
    
    [self callMovieAPI];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return list.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  
    MovieVO *row =  list[indexPath.row];
    MovieCell *cell = (MovieCell *)[tableView dequeueReusableCellWithIdentifier:@"MovieCell"];
    
    if(cell == nil){
        NSLog(@"cell is nill");
        cell = [[MovieCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"MovieCell"];
        cell.accessoryType=UITableViewCellAccessoryDisclosureIndicator;
    }
        
    cell.titleLabel.text = row.title;
    cell.audiAccLabel.text = row.audiAcc;
    cell.openDateLabel.text = row.openDate;
    
    NSLog(@"title : %@",cell.titleLabel.text);
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    NSLog(@"Touch Table Row at %ld", (long)indexPath.row);
}

- (void)callMovieAPI{
    
    AFHTTPSessionManager *manager = [AFHTTPSessionManager manager];
    manager.responseSerializer = [AFHTTPResponseSerializer serializer];
    
    
    NSString *url = @"http://www.kobis.or.kr/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json";
    
    //    apiKey = dd511a8aba40b8ffdfef0fd2ef9f4a1c
    //    Example of API call:
    //     http://www.kobis.or.kr/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json?key=430156241533f1d058c603178cc3ca0e&targetDt=20120101
    
    //    어제 날짜기준. 오늘로 하면 집계가 안되서 에러남.
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyyMMdd"];
    NSString *today = [formatter stringFromDate:[[NSDate date] dateByAddingTimeInterval:-86400.0]];
    
    [manager GET:url
      parameters:@{
                   @"key":@"dd511a8aba40b8ffdfef0fd2ef9f4a1c",
                   @"targetDt":today
                   }
        progress:nil
         success:^(NSURLSessionTask *task, id responseObject) {
             
             NSString *receivedDataString = [[NSString alloc] initWithData:responseObject encoding:NSUTF8StringEncoding];
             NSLog(@"API result : %@",receivedDataString);
             
             NSError *error;
             NSDictionary *apiDictionary = [NSJSONSerialization JSONObjectWithData:responseObject options:kNilOptions error:&error];
             
             
             NSDictionary *boxOfficeResult = (NSDictionary *)[apiDictionary valueForKey:@"boxOfficeResult"];
             NSMutableArray *dailyBoxOfficeList = (NSMutableArray *)[boxOfficeResult valueForKey:@"dailyBoxOfficeList"];
             
             for(NSString *row in dailyBoxOfficeList){
                 mvo = [MovieVO new];
                 
                 mvo.code = (NSString *)[row valueForKey:@"movieCd"];
                 mvo.title = (NSString *)[row valueForKey:@"movieNm"];
                 mvo.audiAcc = [NSString stringWithFormat:@"%@명",(NSString *)[row valueForKey:@"audiAcc"]];
                 mvo.openDate = (NSString *)[row valueForKey:@"openDt"];
                 
                 [list addObject:mvo];
                 
                 [_movieTable reloadData];
             }
             
             
             
         } failure:^(NSURLSessionTask *operation, NSError *error) {
             NSLog(@"%@", error);
         }
     ];
    
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender{
    if([segue.identifier isEqualToString:@"segue_movie_detail"]){
        MovieCell *cell = (MovieCell *)sender;
        
        NSIndexPath *path = [self.movieTable indexPathForCell:cell];
        
        MovieVO *param = list[path.row];
        
        MovieDetailViewController *dvc = (MovieDetailViewController *)[segue destinationViewController];
        
        dvc.mvo = param;
        
        
    }
}

@end

File: MiniProject02/MiniProject02/MovieTableViewController.swift
//
//  MovieTableViewController.swift
//  MiniProject02
//
//  Created by 김성준 on 2016. 11. 29..
//  Copyright © 2016년 sungjun. All rights reserved.
//

import UIKit
import Alamofire
import SwiftyJSON

class MovieTableViewController: UITableViewController {
    
    var list = Array<MovieVO>()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.callKobisAPI(){ movieName in
            print("movieName : \(movieName)")
            
            for row in movieName!{
                self.callNaverAPI(title: row)
            }
        }
    }
    
    func getThumbnailImage(index:Int) -> UIImage{
        let mvo = self.list[index]
        
        if let savedImage = mvo.thumbnailImage{
            return savedImage
        }else{
            let thumbnailURL = URL(string: mvo.thumbnail!)
            let imageData = NSData(contentsOf: thumbnailURL!)
            
            mvo.thumbnailImage = UIImage(data: imageData as! Data)
            
            return mvo.thumbnailImage!
        }
        
    }
    
    func callKobisAPI(completionHandler: @escaping (Array<String>?) -> ()){
        kobisProvider.request(Kobis.showBoxOffice){ result in
            switch result {
            case let .success(response) :
                do{
                    var movieName = Array<String>()
                    let json = try JSON(response.mapJSON())
                    print("Kobis JSON: \(json)")
                    let boxOfficeResult = json["boxOfficeResult"]
                    let dailyBoxOfficeList = boxOfficeResult["dailyBoxOfficeList"]
                    
                    for row in dailyBoxOfficeList{
                        let mvo = MovieVO()
                        
                        mvo.title = row.1["movieNm"].stringValue
                        mvo.openDate = row.1["openDt"].stringValue
                        mvo.audiAcc = row.1["audiAcc"].stringValue
                        
                        movieName.append(mvo.title!)
                        self.list.append(mvo)
                        
                    }
                    
                    completionHandler(movieName)
                }catch{
                    print("error")
                }
            case let .failure(error) :
                print("error : \(error)")
                
            }
        }
    }
    
    func callNaverAPI(title:String){
        var thumbnail : String?
        var rating : Float?
        var link : String?
        
        naverProvider.request(Naver.showMovieInfo(title)){result in
            switch result {
            case let .success(response) :
                do{
                    let json = try JSON(response.mapJSON())
                    print("JSON: \(json)")
                    
                    let item = json["items"][0]
                    
                    thumbnail = item["image"].stringValue == "" ? json["items"][1]["image"].stringValue : item["image"].stringValue
                    rating = item["userRating"].floatValue
                    link = item["link"].stringValue
                    
                    print("thumbnail : \(thumbnail)")
                    print("rating : \(rating)")
                    print("link : \(link)")
                    
                    for mvo in self.list{
                        if mvo.title == title{
                            mvo.thumbnail = thumbnail!
                            mvo.rating = rating!
                            mvo.link = link!
                        }
                    }
                    
                    self.tableView.reloadData()
                    
                }catch{
                    print("error")
                }
            case let .failure(error) :
                print("error : \(error)")
            }
        }
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Table view data source
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return self.list.count
    }
    
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let row = self.list[indexPath.row]
        let cell = tableView.dequeueReusableCell(withIdentifier: "MovieCell") as! MovieCell
        
        //        print("title : \(row.title)")
        
        cell.titleLabel?.text = row.title
        cell.openDateLabel?.text = row.openDate
        if row.rating != nil{
            cell.ratingLabel?.text = "\(row.rating!)"
        }
        cell.audiAccLabel?.text = row.audiAcc
        if row.thumbnail != nil{
            DispatchQueue.main.async {
                cell.thumbnailImageView.image = self.getThumbnailImage(index: indexPath.row)
            }
        }
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        NSLog("Touch Table Row at %d", indexPath.row)
    }
    
    //     MARK: - Navigation
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "segue_movie_detail" {
            let cell = sender as! MovieCell
            
            let path = self.tableView.indexPath(for: cell)
            
            let param = self.list[path!.row]
            
            (segue.destination as? MovieDetailViewController)?.mvo = param
            
            print("segue")
        }
    }
}


